pilopt/src/qse_opt.rs

```diff
@@ -0,0 +1,135 @@
+            let left = try_algebraic_to_quadratic_symbolic_expression(left)?;
+            let right = try_algebraic_to_quadratic_symbolic_expression(right)?;
+            match op {
+                AlgebraicBinaryOperator::Add => left + right,
+                AlgebraicBinaryOperator::Sub => left - right,
+                AlgebraicBinaryOperator::Mul => left * right,
+                AlgebraicBinaryOperator::Pow => {
+                    todo!()
+                }
+            }
+        }
+        AlgebraicExpression::UnaryOperation(AlgebraicUnaryOperation { op, expr }) => match op {
+            AlgebraicUnaryOperator::Minus => -try_algebraic_to_quadratic_symbolic_expression(expr)?,
+        },
+        AlgebraicExpression::PublicReference(_) | AlgebraicExpression::Challenge(_) => return None,
+    })
+}
+
+fn quadratic_symbolic_expression_to_algebraic<T: FieldElement>(
+    expr: &QuadraticSymbolicExpression<T, AlgebraicReference>,
+) -> AlgebraicExpression<T> {
+    let (quadratic, linear, constant) = expr.elements();
+    quadratic
+        .iter()
+        .map(|(l, r)| {
+            quadratic_symbolic_expression_to_algebraic(l)
+                * quadratic_symbolic_expression_to_algebraic(r)
+        })
+        .chain(linear.iter().map(|(v, c)| {
+            AlgebraicExpression::Reference((*v).clone()) * symbolic_expression_to_algebraic(c)
+        }))
+        .chain(iter::once(symbolic_expression_to_algebraic(constant)))
+        .reduce(|acc, x| acc + x)
+        .unwrap()
+}
+
+fn symbolic_expression_to_algebraic<T: FieldElement>(
+    e: &SymbolicExpression<T, AlgebraicReference>,
+) -> AlgebraicExpression<T> {
+    // Problem: div does not work!
+    match e {
+        SymbolicExpression::Concrete(v) => AlgebraicExpression::Number(*v),
+        SymbolicExpression::Symbol(var, _) => AlgebraicExpression::Reference((*var).clone()),
+        SymbolicExpression::BinaryOperation(left, op, right, _) => {
+            let left = Box::new(symbolic_expression_to_algebraic(left));
+            let right = Box::new(symbolic_expression_to_algebraic(right));
+            let op = symbolic_op_to_algebraic(*op);
+            AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation { left, op, right })
+        }
+        SymbolicExpression::UnaryOperation(op, inner, _) => match op {
+            UnaryOperator::Neg => {
+                let expr = Box::new(symbolic_expression_to_algebraic(inner));
+                AlgebraicExpression::UnaryOperation(AlgebraicUnaryOperation {
+                    expr,
+                    op: AlgebraicUnaryOperator::Minus,
+                })
+            }
+        },
+    }
+}
+
+fn symbolic_op_to_algebraic(op: BinaryOperator) -> AlgebraicBinaryOperator {
+    match op {
+        BinaryOperator::Add => AlgebraicBinaryOperator::Add,
+        BinaryOperator::Sub => AlgebraicBinaryOperator::Sub,
+        BinaryOperator::Mul => AlgebraicBinaryOperator::Mul,
+        BinaryOperator::Div => todo!(),
```

Comment:
In the context of this file, this is unreachable, right? Because the QSEs we convert come from `try_algebraic_to_quadratic_symbolic_expression` (which doesn't use division), and then they are simplified by replacing variables by constants, which also doesn't introduce divisions. Correct?

In that case, I'd use `unreachable!()` and put that as a comment.

Reply:
Or, make it `try_quadratic_symbolic_expression_to_algebraic()` and then unwrap.

Reply:
Yes, at the moment it cannot happen.